robot_driver/src/robot_driver.cpp

```diff
@@ -242,54 +247,8 @@ void RobotDriver::mocapCallback(
   Eigen::Vector3d pos;
   quad_utils::pointMsgToEigen(msg->pose.position, pos);
 
-  if (complementary_filter_.low_pass_filter.init) {
-    // Record time diff between messages
-    ros::Time t_now = ros::Time::now();
-    double t_diff_mocap_msg =
-        (msg->header.stamp - last_mocap_msg_->header.stamp).toSec();
-    double t_mocap_ros_latency = (t_now - msg->header.stamp).toSec();
-    last_mocap_time_ = t_now;
-
-    // Apply filter
-    if (abs(t_diff_mocap_msg - 1.0 / mocap_rate_) < mocap_dropout_threshold_) {
-      for (size_t i = 0; i < 3; i++) {
-        // Compute outputs
-        mocap_vel_estimate_(i) =
-            (complementary_filter_.low_pass_filter.C *
-                 complementary_filter_.low_pass_filter.x.at(i) +
-             complementary_filter_.low_pass_filter.D * pos(i))(0, 0);
-
-        // Compute states
-        complementary_filter_.low_pass_filter.x.at(i) =
-            complementary_filter_.low_pass_filter.A *
-                complementary_filter_.low_pass_filter.x.at(i) +
-            complementary_filter_.low_pass_filter.B * pos(i);
-      }
-    } else {
-      ROS_WARN_THROTTLE(
-          0.1,
-          "Mocap time diff exceeds max dropout threshold, hold the last value");
-    }
-  } else {
-    // Init filter, we want to ensure that if the next reading is the same, the
-    // output speed should be zero and the filter state remains the same
-    Eigen::Matrix<double, 3, 2> left;
-    left.topRows(2) =
-        complementary_filter_.low_pass_filter.A - Eigen::Matrix2d::Identity();
-    left.bottomRows(1) = complementary_filter_.low_pass_filter.C;
-
-    Eigen::Matrix<double, 3, 1> right;
-    right.topRows(2) = -complementary_filter_.low_pass_filter.B;
-    right.bottomRows(1) = -complementary_filter_.low_pass_filter.D;
-
-    for (size_t i = 0; i < 3; i++) {
-      complementary_filter_.low_pass_filter.x.at(i) =
-          left.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV)
-              .solve(right * pos(i));
-    }
-
-    complementary_filter_.low_pass_filter.init = true;
-  }
+  comp_filter_estimator_->mocapCallBackHelper(
+      msg, pos, last_mocap_msg_, mocap_rate_, mocap_dropout_threshold_);
```

Comment:
Here is where we would do the dynamic cast (assuming that mocapCallbackHelper is only a member of CF and not just StateEstimation)